llm_demo/orchestrator/langgraph/langgraph_orchestrator.py

```diff
@@ -46,131 +49,125 @@
-            memory=memory,
-            handle_parsing_errors=True,
-            max_iterations=3,
-            early_stopping_method="generate",
-            return_intermediate_steps=True,
-        )
-        agent.agent.llm_chain.prompt = prompt  # type: ignore
-        return UserAgent(client, agent, memory)
-
-    async def close(self):
-        await self.client.close()
-
-    async def invoke(self, prompt: str) -> Dict[str, Any]:
-        try:
-            response = await self.agent.ainvoke({"input": prompt})
-        except Exception as err:
-            raise HTTPException(status_code=500, detail=f"Error invoking agent: {err}")
-        return response
-
-    async def insert_ticket(self, params: str):
-        return await insert_ticket(self.client, params)
-
-    def reset_memory(self, base_message: List[BaseMessage]):
-        self.memory.clear()
-        self.memory.chat_memory = ChatMessageHistory(messages=base_message)
+    is_last_step: IsLastStep
 
 
 class LangGraphOrchestrator(BaseOrchestrator):
-    _user_sessions: Dict[str, UserAgent]
+    _user_sessions: List[str]
     # aiohttp context
     connector = None
 
     def __init__(self):
-        self._user_sessions = {}
+        self._user_sessions = []
 
     @classproperty
     def kind(cls):
         return "langgraph"
 
+    async def create_graph(self, tools, prompt: ChatPromptTemplate, model_name: str):
+        # tool node
+        tool_node = ToolNode(tools)
+
+        # model node
+        model = ChatVertexAI(
+            model=model_name, temperature=0, max_tokens=512, max_iterations=3
+        ).bind_tools(tools)
+        # Add the prompt
+        model_runnable = (lambda messages: prompt + messages) | model
+
+        def call_model(state: UserState, config: RunnableConfig):
+            """
+            Node for function that calls the model
+            """
+            messages = state["messages"]
+            response = model_runnable.invoke(messages, config)
+            if (
+                state["is_last_step"]
+                and isinstance(response, AIMessage)
+                and response.tool_calls
+            ):
+                return {
+                    "messages": [
+                        AIMessage(
+                            id=response.id,
+                            content="Sorry, need more steps to process this request.",
+                        )
+                    ]
+                }
+            return {"messages": [response]}
+
+        def agent_should_continue(self, state: UserState) -> Literal["tools", "end"]:
+            """
+            Function to determine which node is called next
+            """
+            messages = state["messages"]
+            last_message = messages[-1]
+            # If the LLM makes a tool call, then we route to the "tools" node
+            if isinstance(last_message, AIMessage) and last_message.tool_calls:
+                return "tools"
```

Comment:
Should this be a constant?

Reply:
Updated this. The return will be a string, this string will point towards the next node defined in the conditional edges.